exadg/exadg
# A Robin boundary condition for structure module

This PR introduces a general BC of the form
`traction = P*N = - k * displacement - c * velocity - exterior_pressure`.
For this we need several adaptations:

1. `boundary_mass_operator` to shift the discretized remainder of the velocity term to the rhs (similar to the weak damping or general time integration terms, but now with a mass term on the boundary only). This will also come in handy, if we want to project values on the boundary.
2. changes in `operator_base` such that CG discretizations involve some boundary integrals if `evaluate_boundary_integrals() == true` in the fundamental `apply`, `apply_add`, `evaluate`, `evaluate_add` ... operations **and** the remaining MF loops used in MG setup, preconditioner update etc. (building the system matrix or its inverse diagonal)
3. consider the new boundary integral in the linear and nonlinear operators in the structure module
4. scaling factors and updates of the new `BoundaryMassOperator` in `structure/operator.cpp` and the according interface.
5. tests for the new functionality

I have this now in a single PR to thoroughly check if the behavior is as desired (see the tests) and the changes in `OperatorBase` actually work as expected to get good Multigrid convergence. Also, with this you can see the intended use (and extensibility towards Robin coupling conditions in FSI). If need be, I can split into several parts:
1. changes in `OperatorBase`
2. add exterior pressure (Robin condition part 1)
3. add spring (Robin condition part 1)
4. add `BoundaryMassOperator`
5. add dashpot (use `BoundaryMassOperator` and temporal scaling)

Whoever is reviewing this (probably @nfehn :1st_place_medal:), let's have an in-person discussion first!
